### PR TITLE
IFTAQCAB 4 verification data picker options are working (plus DataProvider) - aobitotski

### PR DESCRIPTION
**Pull request by Obitotski**
- added data.conf file for data values and DataProvider interface for reading them 
- have done ticket IFTAQCAB 4 (verification data picker options are working)

### CODE REVIEW

File: src/main/java/com/softserveinc/ita/pageobjects/admin/ProtocolPage.java [hunk 1]
```diff
@@ -1,4 +1,48 @@
 package com.softserveinc.ita.pageobjects.admin;
 
+import com.codeborne.selenide.SelenideElement;
+
+import static com.codeborne.selenide.Selenide.$x;
+import static java.lang.String.format;
+
 public class ProtocolPage extends MainMenu {
+
+    private final SelenideElement startDatePickerButton = $x("(//button[@aria-label='Open calendar'])[1]");
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/admin/ProtocolPage.java:10]: could be as DATE_PICKER_TEMPLATE = "(//button[@aria-label='Open calendar'])[%s]";

File: src/main/java/com/softserveinc/ita/pageobjects/admin/ProtocolPage.java [hunk 2]
```diff
@@ -1,4 +1,48 @@
...
+    private final SelenideElement endDatePickerButton = $x("(//button[@aria-label='Open calendar'])[2]");
+    private final SelenideElement chooseMonthAndYearButton = $x("//button[@aria-label='Choose month and year']");
+    private final SelenideElement searchButton = $x("//button[@type='submit']");
+
+    private final String optionButtonTemplate = "//div[text()='%s']";
+
+    public ProtocolPage chooseStartDate(String startYear, String startMonth, String startDay) {
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/admin/ProtocolPage.java:17]: use LocalDateTime class from org.joda.time lib, and create a class in 'models' package or somewhere else

```
@Builder
@Data
public class DateTimeRange {
    private LocalDateTime start;
    private LocalDateTime end;
}
```

and you would be able to pass only one param into the method (but may not be required within this task)
```
var startDate = LocalDateTime.parse("2016-02-13", forPattern("YYYY-MM-dd"));
var endDate = LocalDateTime.now();
var someRange = DateTimeRange
.builder()
.start(startDate)
.end(endDate)
.build();
```

also, you could create a private method and move some things from your methods:

```
private ProtocolPage chooseDate(LocalDateTime date, int index) {
        $x(format(DATE_PICKER_TEMPLATE, index)).click();
        chooseMonthAndYearButton.click();
        $x(format(optionButtonTemplate, date.getYear())).click();
        $x(format(optionButtonTemplate, date.getMonthOfYear())).click();
        $x(format(optionButtonTemplate, date.getDayOfMonth())).click();

        return this;
    }
```

```
public ProtocolPage chooseStartDate(LocalDateTime date) {
return chooseDate(date, 1);
}

public ProtocolPage chooseEndDate(LocalDateTime date) {
return chooseDate(date, 2);
}
```

File: src/test/java/com/softserveinc/ita/ProtocolTest.java [hunk 3]
```diff
@@ -1,25 +1,49 @@
...
+import static com.softserveinc.ita.pageobjects.util.DataProvider.*;
+import static org.assertj.core.api.Assertions.assertThat;
+
 public class ProtocolTest extends TestRunner {
 
-    @Test
-    public void verifyProtocolPageOpening() {
-        // TODO: 13.05.2022 use config files for credentials and expected url
+    @BeforeMethod
+    public ProtocolPage openProtocolPage() {
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/ProtocolTest.java:17]: in test classes, methods should not return anything, if you want to init ProtocolPage better to create a class variable

```
private ProtocolPage protocolPage;

@BeforeMethod
    public void openProtocolPage() {
        protocolPage = new LoginPage()
                .login(ADMIN_LOGIN, ADMIN_PASSWORD)
                .openProtocolPage();
    }
```

File: src/test/java/com/softserveinc/ita/ProtocolTest.java [hunk 4]
```diff
@@ -1,25 +1,49 @@
...
 
         assertThat(currentUrl)
                 .as("Page url should be " + expectedUrl)
                 .isEqualTo(expectedUrl);
     }
+
+    @Test
+    public void verifyDatePickersOptionsChoosing() {
+
+        var actualResult = new ProtocolPage()
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/ProtocolTest.java:39]:
```
protocolPage
.chooseStartDate
...
```

File: src/test/java/com/softserveinc/ita/ProtocolTest.java [hunk 5]
```diff
@@ -1,25 +1,49 @@
...
-        var currentUrl =  webdriver().object().getCurrentUrl();
+
+        return new ProtocolPage();
+    }
+
+    @Test
+    public void verifyProtocolPageOpening() {
+
+        var expectedUrl = PROTOCOL_PAGE_URL;
+        var currentUrl = webdriver().object().getCurrentUrl();
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/ProtocolTest.java:29]: you could create a util class for such operations with the browser: 'WindowTabHelper'
and create method 
```
public String getCurrentUrl() {
        return getAndCheckWebDriver().getCurrentUrl();
    }
```

**Review by Nikodych — CHANGES_REQUESTED** [review]

**Comment by Obitotski** [comment]
- fixed according to review, except for one moment with code below 

`var startDate = LocalDateTime.parse("2016-02-13", forPattern("YYYY-MM-dd"));
var endDate = LocalDateTime.now();
var someRange = DateTimeRange
.builder()
.start(startDate)
.end(endDate)
.build();`

don`t know where to put it properly

File: src/main/java/com/softserveinc/ita/pageobjects/admin/ProtocolPage.java [hunk 6]
```diff
@@ -2,47 +2,38 @@
...
 
-        startDatePickerButton.click();
+    private ProtocolPage chooseDate(LocalDate date, int index) {
+        $x(format(DATE_PICKER_TEMPLATE, index)).click();
         chooseMonthAndYearButton.click();
-        startYearButton.click();
-        startMonthButton.click();
-        startDayButton.click();
+        $x(format(OPTION_BUTTON_TEMPLATE, date.getYear())).click();
+        $x(format(OPTION_BUTTON_TEMPLATE, date.getMonth().toString().substring(0, 3))).click();
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/admin/ProtocolPage.java:22]: extract into local variable:

```
var month = date
.getMonth()
.toString()
.substring(0, 3);
```

**Review by Nikodych — CHANGES_REQUESTED** [review]

**Pull request merged**